util/global_conn_id.go

```diff
@@ -0,0 +1,453 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"runtime"
+	"sync/atomic"
+
+	"github.com/cznic/mathutil"
+	"github.com/ngaut/sync2"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+type serverIDGetterFn func() uint64
+type connectionIDExistCheckerFn func(connectionID uint64) bool
+
+// ConnectionIDAllocator allocates connection IDs.
+type ConnectionIDAllocator interface {
+	// Init initiates the allocator.
+	Init(serverIDGetter serverIDGetterFn, existedChecker connectionIDExistCheckerFn)
+	// SetServerIDGetter set serverIDGetter to allocator.
+	SetServerIDGetter(serverIDGetter serverIDGetterFn)
+	// NextID returns next connection ID.
+	NextID() uint64
+	// Release releases connectionID to pool.
+	Release(connectionID uint64)
+}
+
+var (
+	_ ConnectionIDAllocator = (*SimpleConnIDAllocator)(nil)
+	_ ConnectionIDAllocator = (*GlobalConnIDAllocator)(nil)
+)
+
+// SimpleConnIDAllocator is a simple auto-increment allocator.
+type SimpleConnIDAllocator struct {
```

Comment:
LocalAutoIncIDAllocator? (which is an ID allocator that simply do auto-increment to allocate ID and will never fail. Wrapping will happen. After that, the uniqueness of the ID will no longer be ensured.)

Reply:
I keep this `SimpleConnIDAllocator`. Please see this [comment](https://github.com/pingcap/tidb/pull/25385#issuecomment-873093138).